ios/Capacitor/Capacitor/KeyValueStore.swift

```diff
@@ -0,0 +1,323 @@
+        get { try? self.get(key) }
+        set {
+            if let newValue {
+                try? self.set(key, value: newValue)
+            } else {
+                try? self.delete(key)
+            }
+        }
+    }
+
+    /// A shared persistent instance of ``KeyValueStore``
+    public static let standard = KeyValueStore(type: .persistent(suiteName: "standard"))
+}
+
+public protocol KeyValueStoreBackend {
+    func `get`<T>(_ key: String, as type: T.Type) throws -> T? where T: Decodable
+    func `set`<T>(_ key: String, value: T) throws where T: Encodable
+    func `delete`(_ key: String) throws
+}
+
+private class FileStore: KeyValueStoreBackend {
+    private let cache = ConcurrentDictionary<Data>()
+    private let decoder = JSONDecoder()
+    private let encoder = JSONEncoder()
+    private let baseUrl: URL
+
+    private init(baseUrl: URL) {
+        self.baseUrl = baseUrl
+    }
+
+    func get<T>(_ key: String, as type: T.Type) throws -> T? where T: Decodable {
+        if let cached = cache[key],
+           let value = try? decoder.decode(type, from: cached) {
+            return value
+        }
+
+        let fileCacheLocation = baseUrl.appendingPathComponent(key)
+        guard FileManager.default.fileExists(atPath: fileCacheLocation.path) else { return nil }
+
+        let data = try Data(contentsOf: fileCacheLocation)
+        let decoded = try decoder.decode(type, from: data)
+
+        cache[key] = data
+        return decoded
+    }
+
+    func set<T>(_ key: String, value: T) throws where T: Encodable {
+        let encoded = try encoder.encode(value)
+        try encoded.write(to: url(for: key), options: .atomic)
+        cache[key] = encoded
+    }
+
+    func delete(_ key: String) throws {
+        cache[key] = nil
+        try FileManager.default.removeItem(at: url(for: key))
+    }
+
+    private func url(for key: String) -> URL {
+        baseUrl.appendingPathComponent(key)
+    }
+
+    private static let instances = ConcurrentDictionary<FileStore>()
+
+    // This ensures we essentially have singletons for accessing file based resources
+    // so we don't have a scenario where two separate instances may be writing to
+    // the same files.
+    static func with(name: String) -> FileStore {
+        let library = try! FileManager
+            .default
+            .url(
+                for: .libraryDirectory,
+                in: .userDomainMask,
+                appropriateFor: nil,
+                create: true
+            )
+
+        let url = library.appendingPathComponent("kvstore").appendingPathComponent(name)
+        if let existing = instances[url.absoluteString] { return existing }
+
+        // Create the folder if it doesn't exist
+        try? FileManager.default.createDirectory(at: url, withIntermediateDirectories: true, attributes: nil)
+
+        let new = FileStore(baseUrl: url)
+        instances[url.absoluteString] = new
```

Comment:
Honestly not sure why I'm keying off the URL instead of just the name. That would allow for checking for its existence before doing all the other url logic.